pallets/phala/src/mining.rs

```diff
@@ -1,58 +1,282 @@
+		StorageMap<_, Twox64Concat, T::AccountId, BalanceOf<T>>;
 
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
-	pub enum Event {
+	pub enum Event<T: Config> {
 		/// Meh. [n]
 		Meh(u32),
+		/// [period]
+		CollingdownExpireChanged(u64),
+		/// [miner]
+		MiningStarted(T::AccountId),
+		/// [miner]
+		MiningStoped(T::AccountId),
+		/// [miner]
+		MiningCleanup(T::AccountId),
+		/// [miner, worker]
+		MinerBounded(T::AccountId, WorkerPublicKey),
+		/// [miner]
+		MinerEnterUnresponsive(T::AccountId),
+		/// [miner]
+		MinerExitUnresponive(T::AccountId),
+		/// [miner, amount]
+		MinerDeposited(T::AccountId, BalanceOf<T>),
+		/// [miner, amount]
+		MinerWithdrawed(T::AccountId, BalanceOf<T>),
 	}
 
 	#[pallet::error]
 	pub enum Error<T> {
-		Meh,
+		BadSender,
+		InvalidMessage,
+		WorkerNotRegistered,
+		GatekeeperNotRegistered,
+		DuplicatedBoundedMiner,
+		BenchmarkMissing,
+		MinerNotFounded,
+		MinerNotReady,
+		MinerIsBusy,
+		CollingdownNotPassed,
+		InsufficientStaking,
 	}
 
 	type BalanceOf<T> =
 		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn sudo_set_collingdownexpire(origin: OriginFor<T>, period: u64) -> DispatchResult {
+			ensure_root(origin)?;
+
+			CollingdownExpire::<T>::mutate(|p| *p = period);
+			Self::deposit_event(Event::<T>::CollingdownExpireChanged(period));
+			Ok(())
+		}
+
+		/// Binding miner with worker
+		///
+		/// Requires:
+		/// 1. Ther worker is alerady registered
+		#[pallet::weight(0)]
+		pub fn bind(
+			origin: OriginFor<T>,
+			miner: T::AccountId,
+			pubkey: WorkerPublicKey,
+		) -> DispatchResult {
+			T::PoolOrigin::ensure_origin(origin)?;
+
+			let now = <T as registry::Config>::UnixTime::now()
+				.as_secs()
+				.saturated_into::<u64>();
+
+			ensure!(
+				registry::Worker::<T>::contains_key(&pubkey),
+				Error::<T>::WorkerNotRegistered
+			);
+
+			ensure!(
+				!MinerBinding::<T>::contains_key(&miner),
+				Error::<T>::DuplicatedBoundedMiner
+			);
+
+			MinerBinding::<T>::insert(&miner, &pubkey);
+			WorkerBinding::<T>::insert(&pubkey, &miner);
+
+			Miners::<T>::insert(
+				&miner,
+				MinerInfo {
+					state: MinerState::Ready,
+					ve: INITIAL_V,
+					v: INITIAL_V,
+					v_updated_at: now,
+					p_instant: 0u64,
+					benchmark: Benchmark {
+						iterations: 0u64,
+						mining_start_time: now,
+					},
+					cooling_down_start: 0u64,
+				},
+			);
+
+			Self::deposit_event(Event::<T>::MinerBounded(miner, pubkey));
+			Ok(())
+		}
+
 		/// Desposits some tokan as stake
 		///
 		/// Requires:
 		/// 1. Ther miner is in Ready state
 		#[pallet::weight(0)]
-		pub fn deposit(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
-			panic!("unimplemented")
+		pub fn deposit(
+			origin: OriginFor<T>,
+			miner: T::AccountId,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			T::PoolOrigin::ensure_origin(origin)?;
```

Comment:
If we only allow a pool to call it, we can move dispatchable functions to a simple function